### PR TITLE
Integrate Snyk into CI

### PR DESCRIPTION
**Pull request by npepinpe**
## Description

This PR adds a new workflow which integrates Snyk with our project. The workflow is executed during CI on every push to `main` or any of the stable branches. It can also be called manually via `workflow_dispatch`.

The workflow can run in two modes: monitor, and test.

> **Note**
> Snyk has two mode of operations. The `test` command will scan the given projects and output a summary of vulnerabilities. The `monitor` command will simply detect the dependencies/vulnerabilities, and upload the snapshot to the Snyk servers. In other words, use `test` to actively report issues with the local code/container image, and `monitor` on every release to update the state of distributed artifacts. Snyk will then notify us of new vulnerabilities based on the uploaded snapshots.

The workflow will test all the distributable artifacts of Zeebe:

- The clients (Go, Java, zbctl)
- All modules listed in our Maven BOM
- The official Docker image, which includes:
  - Scanning the base image
  - Scanning the distribution (i.e. dist module)
  - Scanning zbctl

The workflow fails if it detects [any licenses which violate the organization's license policy](https://docs.snyk.io/scan-application-code/snyk-open-source/licenses), and if it detects any vulnerabilities with a severity of `high` or greater.

If testing, the results are emitted as SARIF files, and uploaded to GitHub Code Scanning. The job summary of the workflow contains links to check them directly. For simplicity, both the PR based link and branch based links are added, regardless of whether the scan occurs during a PR or push.

When monitoring, no SARIF files are emitted.

Currently, the workflow is hooked into our CI workflow to monitor the development head of each supported versions (e.g. `stable/8.2`, `stable/8.1`, etc.), and the latest development head (main).

A final note: the workflow runs on self-hosted, n2-standard-16 instances. This is to speed up building the project (required to scan the Docker image), as running on `ubuntu-latest` is noticeably slower. However, to save cost, we could run on `ubuntu-latest`. Let me know what you think.

## Related issues

related to #9245

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

**Comment by npepinpe** [comment]
You can view the results for the PR via the security tab:

https://github.com/camunda/zeebe/security/code-scanning?query=pr%3A12922+tool%3A%22Snyk+Container%22+is%3Aopen
https://github.com/camunda/zeebe/security/code-scanning?query=pr%3A12922+tool%3A%22Snyk+Open+Source%22+is%3Aopen+

I think normally you could also see it from the checks tab, but that always only shows the latest results, so =/

**Comment by npepinpe** [comment]
Alright, to summarize:

- There's a new workflow (snyk.yml) which is a reusable workflow. Can be called from other workflows, or directly via `workflow_dispatch`.
- The workflow can test or monitor projects.
  - When monitoring, no SARIF output is produced, and there is no job summary either.
- When testing, SARIF files are uploaded to GitHub Code Scanning and accessible there. You can filter by PR, branch, etc. The contents are divided into two tools: Snyk Open Source (for Maven, Go, etc.), and Snyk Container (for the Docker image).
- The workflow will fail if there are high or critical severity vulnerabilities.
- The workflow will fail if there "high" severity licenses (as defined [here](https://docs.snyk.io/scan-application-code/snyk-open-source/licenses)). This can be configured via license policies.
- The projects tested/monitored are those we distribute: the Docker image, the BOM, all modules included in the BOM, and the Go clients.
- To avoid duplicated dependencies when scanning the container image, [we exclude application vulnerabilities](https://docs.snyk.io/scan-containers/using-snyk-container/detecting-application-vulnerabilities-in-container-images). This could be an issue in the future if we add more things to our container that aren't tracked through the other modules.

**Comment by npepinpe** [comment]
@abbasadel - I added you as I think you're interested in this, but in case you're busy, no need for 2 people to review I think.

**Comment by npepinpe** [comment]
@megglos / @abbasadel - let's put a pin on this PR for a bit. I'll reuse this as a base for the monitor portion which will be included in our release process. However, it occurred to me that this will fail PRs when a dependency has a high/critical vulnerability, even if the PR is not introducing it.

So I would propose moving on with the monitor part, but hold off on the testing for now, and rely on the GitHub integration.

**Comment by npepinpe** [comment]
Here's a successful monitor run: https://github.com/camunda/zeebe/actions/runs/5190473997/jobs/9357369888

You can see in Snyk all the projects under the `np-snyk` grouping. I'll delete them later.

**Comment by npepinpe** [comment]
One more learning, it seems the `snyk container` command doesn't let you specify a target reference, meaning all Docker projects are put under the same reference, being the image name. This means all Docker scanned projects are pushed under the `camunda/zeebe` reference :shrug: Project names follow the format `zeebe(BRANCH):project`, so at least the name should indicate the version it references.

**Comment by npepinpe** [comment]
Looks like this:

image

It also seems to put the Docker projects in a separate CLI section and flag them as "private" targets :shrug: This is just par for the course with Snyk UX from my side

File: .github/workflows/ci.yml
Comment (megglos) [inline .github/workflows/ci.yml:557]: ❌ `github.ref` starts with `refs/head/<branch_name>`, see also how we match for main in line 570 . An alternative might be `github.ref_name` which only contains the plain branch name

File: .github/workflows/ci.yml
Comment (megglos) [inline .github/workflows/ci.yml:552]: ❓ this only applies to stable/main pushes, the job for PRs is missing?

**Review by megglos — APPROVED** [review]
Nice! Looking forward to main and stable dev states being monitored going forward!

> The workflow is executed during CI on every push and pull request

If I didn't miss anything Pull request runs are not yet configured in the ci workflow? See comment on `ci.yml`. I guess we could enable that with a follow-up as well though.

Besides some questions just one potential issue on the ref matching condition, if I understood the content of the ref property correctly, wouldn't require another review though.

**Comment by npepinpe** [comment]
The pull request thing is a leftover from when I wrote this PR specifically for testing every pull :sweat_smile:

File: .github/workflows/ci.yml
Comment (npepinpe) [reply .github/workflows/ci.yml:552]: There shouldn't be, I'll amend the description

File: .github/workflows/snyk.yml
Comment (npepinpe) [reply .github/workflows/snyk.yml:21]: Correct, only for container scan. I have the build param there for if you want to call the workflow from another job where the artifacts are already all built.

File: .github/workflows/snyk.yml
Comment (megglos) [reply .github/workflows/snyk.yml:67]: got it, https://github.com/camunda/zeebe/pull/12922#discussion_r1221187735 answers this then

Comment (npepinpe) [reply .github/workflows/snyk.yml:67]: The build parameter is more for integration with other workflows :thinking: We can revisit this when we integrate it with the release workflow itself. Possibly we will always checkout and always build anyway.

Comment (megglos) [reply .github/workflows/snyk.yml:67]: as this workflow runs an own job, will the workspace of the caller actually be shared? 🤔

File: .github/workflows/ci.yml
Comment (npepinpe) [reply .github/workflows/ci.yml:557]: Good catch :+1:

Comment (npepinpe) [reply .github/workflows/ci.yml]: I also added a check that we do it on push. Not that we do PRs for main/stable, but just in case.

File: .github/workflows/snyk.yml
Comment (npepinpe) [reply .github/workflows/snyk.yml:67]: Good point. We can keep the build for when the Docker image has been pushed (i.e. after release), but then we should always check out.

**Comment by npepinpe** [comment]
Please a quick double check of the last 2 commits :pray:

**Review by megglos — APPROVED** [review]
LGTM! 🤞

**Comment by npepinpe** [comment]
bors merge

**Pull request merged**
